### PR TITLE
ADB is affected to cve-2020-25695 (sandbox escape)

### PR DESCRIPTION
**Pull request by darthunix**
CVE-2020-25695 exploits vulnerability for enqueued security-restricted operations at commit code. Specifically defered triggers on index relations, materialize view refresh and cursors with hold are affected. An attacker having permission to create non-temp objects in at least one schema can execute arbitrary SQL functions under the identity of the bootstrap superuser.

Current PR consists of two commits. The first one demonstrates the problem in GPDB with analyze and defered triggers. The second one is a backport of https://github.com/postgres/postgres/commit/43ebfea5a988582e9edc752cb1e22e929edac03b with an important addition - we also add security context flags to dispatch message protocol. Otherwise QD backend can have security flags in a static memory, but QEs executing some trigger function don't know anything about these flags and can't cancel a restricted operation.

### CODE REVIEW

**Comment by x4m** [comment]
Hi! Thanks for the backport of fix for an important security issue.
BTW maybe [CVE-2020-14350](https://github.com/postgres/postgres/commit/7eeb1d9861b0a3f453f8b31c7648396cdd7f1e59) worth porting as well?

**Comment by darthunix** [comment]
@x4m thanks, `CVE-2020-14350` would be also added to our backlog.

File: src/test/regress/sql/privileges.sql [hunk 1]
```diff
@@ -1068,6 +1068,88 @@ set session role regressuser1;
...
+
+-- Execute a trap with superuser.
+RESET SESSION AUTHORIZATION;
+
+ANALYZE trap;
+
+-- Check "regress_sro_user" superuser privileges on segments.
+SELECT DISTINCT is_superuser_on_segments('regress_sro_user');
+
+DROP TABLE trap;
```
Comment (x4m) [inline src/test/regress/sql/privileges.sql:1144]: This line SIGSEGVs on segments sometimes.

Comment (darthunix) [reply src/test/regress/sql/privileges.sql:1144]: @x4m do you have a core dump or a stack trace? I can't reproduce SIGSERV on my laptop.

Comment (darthunix) [reply src/test/regress/sql/privileges.sql:1144]: I've fixed regression test output (it was incorrect in a draft) and modified security context cleanup on QEs. Now all regression and unit tests pass for Postgres planner and ORCA optimizer. Can you recheck the SIGSERV?

Comment (x4m) [reply src/test/regress/sql/privileges.sql:1144]: In my environments fix works fine.

**Review by x4m — APPROVED** [review]

**Review by leskin-in — APPROVED** [review]

**Comment by darthunix** [comment]
Opened a PR to master - https://github.com/greenplum-db/gpdb/pull/12370

**Review by InnerLife0 — APPROVED** [review]

**Comment by Stolb27** [comment]
@darthunix 
You have modified `cdbdisp_buildUtilityQueryParms`, `cdbdisp_buildPlanQueryParms` calls to pass context to segments. But what about `cdbdisp_buildCommandQueryParms` and `qdSerializeDtxContextInfo` calls? The first of them is used in `gp_execute_on_server` wrapped by SQL. So it seems that user can replace mpp insert to gp_execute_on_server('insert into') in the index fundtion. Please check this statement.

**Comment by darthunix** [comment]
@Stolb27 thanks for this catch.
1. `gp_execute_on_server()` requires `regress.so` dynamic library that should be built and installed to production cluster. Do we distribute `regress.so` in our builds? Moreover, superuser should create `gp_execute_on_server()` SQL function manually as a regular user can't create a C function.
2. Current PR doesn't protect superuser from manually running some malware security definer function. Instead it give protection only from running maintenance operations that can invisibly execute some malware code. So if a superuser executes insert into executor table he still gives superuser to regress_sro_user after this fix. Superusers should be careful.
3. We should care about dispatch code only when we send some query parts to segments (slices or query text). Anyway, `gp_execute_on_server()` [fails](https://github.com/arenadata/gpdb/blob/524661e88a74ef9da42fa8a1517680a5ad5ecbb7/src/test/regress/regress_gp.c#L2040) with an error if it is executed not on coordinator. In a case of ANALYZE we don't have any problems as non-system tables don't run ANALYZE code on coordinator. BUT we can still be in trouble with matview refresh and I plan to check this case tomorrow.

**Comment by darthunix** [comment]
@Stolb27 I've failed to break this patch with `gp_execute_on_server()` on materialized view refresh (analyze on a table with malware index function is not affected due to on segment execution). The problem is that is we create a materialized view from a function
```sql
CREATE or replace FUNCTION mv_action() RETURNS bool LANGUAGE sql SECURITY INVOKER AS
$$select gp_execute_on_server(
    0,
    'SET allow_segment_DML = ON; DECLARE c CURSOR WITH HOLD FOR SELECT unwanted_grant()'
); SELECT true$$;
```
we do not benefit from CURSOR WITH HOLD delayed user privileges check - we get an unprivileged user on segments or on entrydb (if segno -1). If we rewrite the function to
```sql
CREATE or replace FUNCTION mv_action() RETURNS bool LANGUAGE sql SECURITY INVOKER AS
$$SET allow_segment_DML = ON; DECLARE c CURSOR WITH HOLD FOR SELECT gp_execute_on_server(
    0,
    'select unwanted_grant();'
); SELECT true$$;
``` 
our patch works and we restrict cursor WITH HOLD within security-restricted operation.

I think we should not take into account `gp_execute_on_server()` anyway as it is a very specific C function from regression dynamic library that is not added to our release.

**Comment by darthunix** [comment]
Hmmm... @Stolb27 was right - it is possible to break our fix if we use `gp_execute_on_server()` (though we should never deal with it anywhere except regression tests in CI).
```sql
CREATE OR REPLACE FUNCTION trap_sfunc(integer) RETURNS integer
LANGUAGE sql SECURITY INVOKER AS
$_$SELECT gp_execute_on_server(0, 'SET allow_segment_DML = ON; RESET SESSION AUTHORIZATION; INSERT INTO executor VALUES (current_user);'); SELECT $1$_$;
```
But I still vote that this case should not be handled in the patch.

**Pull request merged**